### PR TITLE
Queen - Piece specific move logic code

### PR DESCRIPTION
**Pull request by brantwellman**
- Queen model logic
  - is_move_allowed? method completed
  - legit_moves method completed
- Test
  - tests in piece_test.rb

### CODE REVIEW

**Comment by kenmazaika** [comment]
You guys are coding machines! :)

File: app/models/queen.rb [hunk 1]
```diff
@@ -1,16 +1,59 @@
 class Queen < Piece
 
 	def is_move_allowed?(new_x, new_y)
-		# return a boolean (true if move is valid, else false)
-		# most pieces have two steps: 
-		# (1) is the target move allowed by the game piece logic? 
-		# (2) is there an obstruction in the way? the game model
-		#  	  has a function that checks this:
-		#  		   is_move_obstructed?(piece_id, new_x, new_y)
+		# check whether suggested move obeys game logic
```
Comment (kenmazaika) [inline app/models/queen.rb]: Same note about the copy/pasted/sameness of this code.  Probably want to move this into the piece model (see my comment to nan)

File: app/models/queen.rb [hunk 2]
```diff
@@ -1,16 +1,59 @@
...
+			move_logic_is_valid = true
+		end		
+
+		# if the destination location obeys the game rules, then check for obstructions
+		if move_logic_is_valid
+			# if is_move_obstructed returns true, then return 'false' for main function,
+			# and vice versa
+			return !self.game.is_move_obstructed?(self.id, new_x, new_y)
+		else 
+			return nil
```
Comment (kenmazaika) [inline app/models/queen.rb]: Again, return `false` here, not `nil`

File: app/models/queen.rb [hunk 3]
```diff
@@ -1,16 +1,59 @@
...
+			move6 = [x_move2, y_move1]
+			# Diagonal to bottom right
+			move7 = [x_move1, y_move2]
+			# Diagonal to bottom left
+			move8 = [x_move2, y_move2]
+
+			moves.push(move1, move2, move3, move4, move5, move6, move7, move8)
+		end
+		
+		moves.delete_if {|moves| moves.include?(nil)}
```
Comment (kenmazaika) [inline app/models/queen.rb]: `flatten` :)

File: app/models/queen.rb [hunk 4]
```diff
@@ -1,16 +1,59 @@
...
 	end
 
 	def legit_moves
 		# return an array of squares that the piece can move to
+		x_init = self.x_coord
+		y_init = self.y_coord
+		tiles = [1, 2, 3, 4, 5, 6, 7]
+		moves = []
+		
+		tiles.each do |x|
```
Comment (kenmazaika) [inline app/models/queen.rb]: A lot of this code is copy/pasted from the rook/bishop.  What do you think about making a method called

```
def legit_moves_for_rook
```

and

```
def legit_moves_for_bishop
```

into the piece model.  In the rook and the bishop we could just delegate the call to the correct method in the piece model.  The magic here is that we could implement this like:

```
def legit_moves
  legit_moves_for_rook + legit_moves_for_bishop
end
```

File: test/models/piece_test.rb [hunk 5]
```diff
@@ -75,4 +75,25 @@ class PieceTest < ActiveSupport::TestCase
 
 	end
 
+	test "check queen moves are legit" do
+		piece = FactoryGirl.create(:queen, :x_coord => 2, :y_coord => 2, :color => "white")
```
Comment (kenmazaika) [inline test/models/piece_test.rb]: Again, test obstructed case.

**Comment by kenmazaika** [comment]
Awesome work guys.  You're crushing this back-end logic! :)

File: app/models/queen.rb (hunk 1 above)
Comment (brantwellman) [inline app/models/queen.rb]: I've opened up an Issue so that this can be dealt with separately in a separate pull request for all of the Piece_type models.

File: app/models/queen.rb (hunk 2 above)
Comment (brantwellman) [inline app/models/queen.rb]: Got it. That will be updated.

File: app/models/queen.rb (hunk 3 above)
Comment (brantwellman) [inline app/models/queen.rb]: We tried `flatten` and `compact`, neither of which worked.
Turns out we were mislead.  ;-)
We are going to keep the `delete_if` statement for each of these piece specific model methods

**Comment by brantwellman** [comment]
Please don't merge this pull request just yet. I will be merging the rook-move branch into this one, plus I need to work on a few other details before we can actually merge it.

**Comment by bobbyali** [comment]
That reminds me - is there a way to mark a pull request to prevent a merge?
Imagine the following scenario:

- I create a pull request that I think is ready, and passes the tests.
- Ken tells me that it could do with a big tweak.
- I then want to make sure that nobody else merges it before I finish the
tweak.

Any thoughts?

Rehan

**Comment by kenmazaika** [comment]
Typically you'll close the pull request, then re-open it later.

**Comment by bobbyali** [comment]
If you close a pull request, can you still comment on it?

**Comment by kenmazaika** [comment]
@bobbyali, yup!

**Comment by brantwellman** [comment]
Lesson learned! Next time I'll close the pull request and open a new one when it's ready to be merged.

**Comment by brantwellman** [comment]
This branch should be ready to be merged now. It includes code for Queen AND Rook moves along with tests and other changes:

- Queen model logic
  - is_move_allowed? method completed
  - legit_moves method completed
- Test
  - additional tests in piece_test.rb

- Rook model logic
  - is_move_allowed? method completed
  - legit_moves method completed
- Test
  - additional tests in piece_test.rb

- Edits to is_move_obstructed? to remove bugs

**Comment by kenmazaika** [comment]
Awesome! :)

**Comment by brantwellman** [comment]
Looks like there are conflicts with my pull request.... :(

Now should I be the one that does the merge and deals with the actual conflicts? And should I wait to do that until someone (in this case, Erin), reviews the code and gives it the :+1:

**Comment by kenmazaika** [comment]
@brantwellman, usually the person who is proposing the change will resolve the conflict first.  Resolving it prior to being code reviewed is best, because that way changes that happen as a result of the merge conflict will get reviewed.

**Comment by brantwellman** [comment]
Got it! I'll get to it as soon as I can...probably an hour or two.

**Comment by brantwellman** [comment]
Good to go!

**Comment by kenmazaika** [comment]
@brantwellman's crazy code is finally in master.  Good work, Brant.  That was a good one :)

**Comment by nannigan** [comment]
Just so happened I saw mention of WebRTC in info about a conference. Whassat?
I see on http://www.webrtc.org/
WebRTC is a free, open project that provides browsers and mobile applications with Real-Time Communications (RTC) capabilities via simple APIs. The WebRTC components have been optimized to best serve this purpose. 

Will that supersede the usefulness of something like Firebase?
Or does the db part of Firebase make it more powerful?
Or?

Nan

**Comment by kenmazaika** [comment]
WebRTC is the technology that Google Hangouts uses.  It is a protocol that
many browsers use to facilitate the realtime streaming of video (I believe
other data can be streamed as well as video, but it is largely video due to
the volume of data).  I believe OpenTok <https://tokbox.com/> is the
"Firebase" of WebRTC.  I've played around with OpenTok a little and it's
pretty cool.

Firebase uses the underlying technology of WebSockets, which is used for
realtime streaming of text/json data.

WebRTC is pretty interesting, but as far as I know building out the server
components is quite difficult.
